File: c_plus_plus/leetcode_1046.cpp
#include <iostream>
#include <queue>
using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int> queue;
        for(auto s: stones) {
            queue.push(s);
        }
        while(queue.size() > 1){
            int a = queue.top();
            queue.pop();
            int b = queue.top();
            queue.pop();
            queue.push(abs(a - b));
        }

        if(queue.size() == 1) {
            return queue.top();
        } else {
            return 0;
        }
    }
};

int main() {
    Solution s = Solution();
    vector<int> stones = {2,7,4,1,8,1};
    int last_weight = s.lastStoneWeight(stones);
    cout << last_weight << endl;

}

File: py/leetcode_1046.py
import heapq
class Solution(object):
    def lastStoneWeight(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        stones = [-s for s in stones]
        heapq.heapify(stones)
        while len(stones) > 1:
            heapq.heappush(stones, -abs(heapq.heappop(stones) - heapq.heappop(stones)))

        if len(stones) == 1:
            return -heapq.heappop(stones)
        else:
            return 0

if __name__ == '__main__':
    print Solution().lastStoneWeight([2,7,4,1,8,1])

